cmd/ebrelayer/relayer/prophecyHandler.go

```diff
@@ -0,0 +1,264 @@
+
+			if !sub.handleBatchProphecyCompleted(prophecyInfoMap, batchStartNonce, batchEndNonce) {
+				sub.SugaredLogger.Errorw("fail to process prophecy.")
+				return
+			}
+			return
+		}
+		batchEndNonce = prophecy.GlobalNonce
+		if batchEndNonce > batchStartNonce+5 {
+			if !sub.handleBatchProphecyCompleted(prophecyInfoMap, batchStartNonce, batchEndNonce) {
+				sub.SugaredLogger.Errorw("fail to process prophecy.")
+				return
+			}
+			batchStartNonce = batchEndNonce + 1
+		}
+		nonce++
+	}
+
+	sub.handleBatchProphecyCompleted(prophecyInfoMap, batchStartNonce, batchEndNonce)
+}
+
+// Parses event data from the msg, event, builds a new ProphecyClaim, and relays it to Ethereum
+func (sub CosmosSub) handleBatchProphecyCompleted(
+	prophecyInfoMap map[uint64]types.ProphecyInfo,
+	batchStartNonce uint64,
+	batchEndNonce uint64) bool {
+
+	sub.SugaredLogger.Infow(
+		"handle batch prophecy completed.",
+		"cosmosMsg", prophecyInfoMap,
+	)
+
+	var batchProphecyInfo []types.ProphecyInfo
+	for batchStartNonce <= batchEndNonce {
+		batchProphecyInfo = append(batchProphecyInfo, prophecyInfoMap[batchStartNonce])
+		batchStartNonce++
+	}
+
+	client, auth, target, err := tryInitRelayConfig(sub)
+	if err != nil {
+		sub.SugaredLogger.Errorw("failed in init relay config.",
+			errorMessageKey, err.Error())
+		return false
+	}
+
+	// Initialize CosmosBridge instance
+	cosmosBridgeInstance, err := cosmosbridge.NewCosmosBridge(target, client)
+	if err != nil {
+		sub.SugaredLogger.Errorw("failed to get cosmosBridge instance.",
+			errorMessageKey, err.Error())
+		return false
+	}
+
+	maxRetries := 5
+	i := 0
+
+	for i < maxRetries {
+		err = txs.RelayBatchProphecyCompletedToEthereum(
+			batchProphecyInfo,
+			sub.SugaredLogger,
+			client,
+			auth,
+			cosmosBridgeInstance,
+		)
+
+		if err != nil {
+			sub.SugaredLogger.Errorw(
+				"failed to send new prophecy completed to ethereum",
+				errorMessageKey, err.Error(),
+			)
+		} else {
+			break
+		}
+		i++
+	}
+
+	if i == maxRetries {
+		sub.SugaredLogger.Errorw(
+			"failed to broadcast transaction after 5 attempts",
+			errorMessageKey, err.Error(),
+		)
+		return false
+	}
+
+	// process successfully complete
+	return true
+
+}
+
+func (sub CosmosSub) getUnhandledProphecies(client *tmClient.HTTP, currentBlock uint64, lastSubmittedNonce uint64) (uint64, map[uint64]types.ProphecyInfo) {
+	// return variable
+	prophecyInfoMap := map[uint64]types.ProphecyInfo{}
+	maxGlobalNonce := uint64(0)
+
+	// determine the last possible block we will scan according to prophecy life time
+	var endBlock uint64
+	if currentBlock > ProphecyLifeTime {
+		endBlock = currentBlock - ProphecyLifeTime
+
+	} else {
+		endBlock = 0
+	}
+
+	ctx := context.Background()
+	// if found the next one
+	found := false
+	for currentBlock > endBlock {
+		// not go to priviouos block if found.
+		if found {
+			break
+		}
+		tmpBlock := int64(currentBlock)
+		block, err := client.BlockResults(ctx, &tmpBlock)
```

Comment:
Let's update this to use the new API to gather a whole bunch of data at once from sifnode instead of querying block by block.

Reply:
implemented.